packages/angular/build/src/tools/angular/transformers/lazy-routes-transformer.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import assert from 'node:assert';
+import { relative } from 'node:path/posix';
+import ts from 'typescript';
+
+export function lazyRoutesTransformer(
+  compilerOptions: ts.CompilerOptions,
+  compilerHost: ts.CompilerHost,
+): ts.TransformerFactory<ts.SourceFile> {
+  const moduleResolutionCache = compilerHost.getModuleResolutionCache?.();
+  assert(
+    typeof compilerOptions.basePath === 'string',
+    'compilerOptions.basePath should be a string.',
+  );
+  const basePath = compilerOptions.basePath;
+
+  return (context: ts.TransformationContext) => {
+    const factory = context.factory;
+
+    const visitor = (node: ts.Node): ts.Node => {
+      if (!ts.isObjectLiteralExpression(node)) {
+        return ts.visitEachChild(node, visitor, context);
+      }
+
+      let hasPathProperty = false;
+      let loadComponentOrChildrenProperty: ts.PropertyAssignment | undefined;
+      for (const prop of node.properties) {
+        if (!ts.isPropertyAssignment(prop) || !ts.isIdentifier(prop.name)) {
+          continue;
+        }
+
+        const propertyNameText = prop.name.text;
+        if (propertyNameText === 'path') {
+          hasPathProperty = true;
+        } else if (propertyNameText === 'loadComponent' || propertyNameText === 'loadChildren') {
+          loadComponentOrChildrenProperty = prop;
+        }
+
+        if (hasPathProperty && loadComponentOrChildrenProperty) {
+          break;
+        }
+      }
+
+      const initializer = loadComponentOrChildrenProperty?.initializer;
+      if (
+        !hasPathProperty ||
+        !initializer ||
+        !ts.isArrowFunction(initializer) ||
+        !ts.isCallExpression(initializer.body) ||
+        !ts.isPropertyAccessExpression(initializer.body.expression) ||
+        initializer.body.expression.name.text !== 'then' ||
+        !ts.isCallExpression(initializer.body.expression.expression) ||
+        initializer.body.expression.expression.expression.kind !== ts.SyntaxKind.ImportKeyword
+      ) {
+        return ts.visitEachChild(node, visitor, context);
+      }
```

Comment:
There are a few issues with this approach:  
1. The dependency graph will change, making `./some-dep` a dependency of the `routes` files rather than being limited to `./my-route`. This could lead to the bundler splitting files differently, potentially impacting the build structure.  
2. The `import()` function does not preload the JavaScript; instead, it loads and executes it immediately, which is not desired behavior.

Reply:
Wouldn't this still give you two output chunks? I imagine the optimized build will look like:

```typescript
{
  path: '',
  loadComponent: async () => {
    import('./chunk-abc123.js'); // Pre-load a dependency we know we'll need.
    const MyRoute = await import('./chunk-def456').MyRoute;
    return MyRoute;
  },
  ...(typeof ngServerMode !== "undefined" && ngServerMode ? { ɵentryNames: ['src/app/some-dep.ts', 'src/app/my-route.ts'] }: {})
}
```

Does esbuild try to join those two dynamic imports or something?

Also I'm using the term "preload" kind of loosely. I get that's going to execute `./chunk-abc123.js` ASAP. The example is a bit contrived, but just calling out that other dynamic imports could be happening.

A simpler example might be:

```typescript
{
  // ...
  loadComponent: () => {
    logger.log('Loading route!');
    return import('./my-route');
  }
}
```

It's definitely uncommon to do other work in these load functions, but I don't think that should necessarily exclude them from this optimization.

Reply:
Whilst, this could indeed happen this to me is quite unusual. I think we could either handle this in the future if this comes up, or just say it’s an unsupported case.